dbconn/factory.go

```diff
@@ -122,15 +123,58 @@ func (f factory) Make(cfg blip.ConfigMonitor) (*sql.DB, string, error) {
 		blip.Debug("%s: auto-enabled skip-verify on socket with TLS but no hostname", cfg.MonitorId)
 	}
 
-	// Load and register TLS, if any
+	// Load and register TLS, if any, and generate a hash of the TLS certs
+	// so we can check to see if they changed.
+	// In the event that we need to check that our credentials have changed we will
+	// also need to verify that the certificates have not also been changed.
 	tlsConfig, err := cfg.TLS.LoadTLS(portSuffix.ReplaceAllString(cfg.Hostname, ""))
 	if err != nil {
 		return nil, "", err
 	}
+
 	if tlsConfig != nil {
 		mysql.RegisterTLSConfig(cfg.MonitorId, tlsConfig)
 		params = append(params, "tls="+cfg.MonitorId)
 		blip.Debug("%s: TLS enabled", cfg.MonitorId)
+
+		// TLS is configured, so make sure we reload it when the credentials are reloaded in case
+		// it was changed
+		origCredentialFunc := credentialFunc
+		credentialFunc = func(ctx context.Context) (Credentials, error) {
+			creds, err := origCredentialFunc(ctx)
+
+			if err != nil {
+				return creds, err
+			}
+
+			// Reload from the original configuration by default
+			var tls blip.ConfigTLS = cfg.TLS
+
+			// If the credentials function returned back a config, use that instead
+			if creds.TLS.Set() {
+				tls = creds.TLS
```

Comment:
This is too confusing. The confusion is `config.tls` (or `config.monitors.tls`) vs. `Credentials.TLS`: which one's the "real" TLS and why might one be set by not the other? (So I'd call this a case of DRY violation.) I think the latter only comes from a my.cnf? But at some point the code needs to DRY and make 1 TLS the authoritative one.